src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -1573,6 +1656,39 @@ public int compare(List<ByteBuffer> a, List<ByteBuffer> b)
         }
     }
 
+    private static class IndexColumnComparator<T> extends ColumnComparator<List<ByteBuffer>>
+    {
+        private final SingleRestriction restriction;
+        private final int columnIndex;
+
+        // VSTODO maybe cache in prepared statement
+        public IndexColumnComparator(SingleRestriction restriction, int columnIndex)
+        {
+            this.restriction = restriction;
+            this.columnIndex = columnIndex;
+        }
+
+        @Override
+        public boolean indexOrdering()
+        {
+            return true;
+        }
+
+        @Override
+        public Comparator<List<ByteBuffer>> prepareFor(TableMetadata table, QueryOptions options)
+        {
+            Index index = restriction.findSupportingIndex(IndexRegistry.obtain(table));
```

Comment:
This finds the first of those indexes that support the restriction. However, the index selection for the query is made by `SIM#getBestIndexQueryPlanFor`. So I think a query using an index implementation could end up using the comparator provided by another index implementation. This doesn't happen in practice because SAI happens to be the only provided implementation using sorting, albeit it could still happen with custom implementations.

Reply:
I agree. I have added a `findSupportingIndexFromQueryPlan` to `Restriction` and am using that here instead. While it is unlikely that this is a problem, it is a good idea to have the index lookups consistent.